### PR TITLE
Update data source config schema

### PR DESCRIPTION
**Pull request by pritipath**
Issue Number: [172](https://github.com/Green-Software-Foundation/carbon-aware-sdk/issues/172)

# DEPENDENT PR
> **Warning**
> This PR depends on commits in other PR
> **DO NOT** merge this PR until  #191 merged and this branch is updated against the latest dev branch

## Summary
Provides the ability to configure data source interfaces with a different data source implementation.   

## Changes

- List of comprehensive changes

## Checklist

- [ ] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [ ] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?
If yes, what are the expected API Changes? Please link to an API-Comparison workflow with the API Diff.

## Is this a breaking change? Yes
If yes, what workflow does this break? 
Application startup will fail if data source configuration does not represent the desired schema 

## Anything else?
Other comments, collaborators, etc.
> Please follow [GitHub's suggested syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to link Pull Requests to Issues via keywords

This PR Closes #<issue_number>
https://github.com/Green-Software-Foundation/carbon-aware-sdk/issues/172

### CODE REVIEW

**Review by JenMadiedo — APPROVED** [review]
Lot's of work here, but really cleans up how the data sources are configured, thanks!

File: docs/architecture/aggregators.md [hunk 1]
```diff
@@ -2,10 +2,10 @@
...
+Each aggregator can support a wide variety of consumer requests whose arguments may be required to access data from one or more data sources. The input to the aggregator must be generic enough to handle those cases, but specific enough to allow enforcement of required fields and validations (i.e., a list field cannot be empty, a time field cannot be in the past etc.). The `CarbonAwareParameters` class handles these concerns for both `EmissionsAggregator` and `ForecastAggregator`. Each public method in the aggregator receives an instance of this "Parameters" class.  Future aggregators will create their own "Parameters" class to manage their argument needs.
 
 ## Carbon Aware Parameters
 The `CarbonAwareParameters` class allows the user to pass in a unique parameter instance to the public `CarbonAwareAggregator` methods with the specific parameters needed by that call.
```
Comment (JenMadiedo) [inline docs/architecture/aggregators.md:11]: Missed a reference to `CarbonAwareAggregator` here

File: docs/architecture/data-sources.md [hunk 2]
```diff
@@ -32,14 +32,14 @@ dotnet add package Microsoft.Extensions.DependencyInjection
 ### Adding/Extending a Data Source Interface
 Each new data source should extend from a generic data source interface. A data source interface defines all the parameters and functions that any data source that falls under it's purview must define/implement. By defining the interface, it allows the SDK to switch between the set of data sources seamlessly because they all share the same input functions and output types. 
 
-Currently there is one data source interface defined -  `ICarbonIntensityDataSource` - which is the interface for all data sources that provide carbon intensity data. A new data source interface should be defined only when there is a new general area of calculation that is being introduced to the SDK.
```
Comment (JenMadiedo) [inline docs/architecture/data-sources.md:35]: the second part of the sentence doesn't quite make sense anymore. Maybe something along the lines of " - which cover data sources that provide carbon intensity data" or something like that

File: docs/architecture/overview.md [hunk 3]
```diff
@@ -24,16 +24,18 @@ Aggregators have knowledge of the different data source interfaces in the data t
 
 For example, the `Carbon Intensity Aggregator` handles requests for various carbon emissions information. It can calculate the average carbon emissions over a time period, or the best carbon emissions given a set of locations. It can transform forecasted carbon emissions to suit particular use-cases. It can also just deliver the emissions data points in a standard schema without performing any calculations.
```
Comment (JenMadiedo) [inline docs/architecture/overview.md:25]: Another missed reference to the `Carbon Intensity Aggregator`. This whole paragraph might need a quick rewriting

File: docs/configuration.md [hunk 4]
```diff
@@ -40,26 +61,276 @@ cd src/CarbonAware.WebApi
...
+        "Proxy": {
+          "useProxy": true,
+          "url": "http://10.10.10.1",
+          "username": "proxyUsername",
+          "password": "proxyPassword"
+        }
+      },
+      "Json": {
+        "Type": "Json",
+        "DataFileLocation": "test-data.json"
```
Comment (esgraham) [inline docs/configuration.md:93]: The file name needs to be changed to test-data-azure-emissions.json here and in the rest of the document.

**Review by esgraham — CHANGES_REQUESTED** [review]
Please update the filename in the configuration.md file.  The other comments are formatting suggestions.

File: src/CarbonAware.WebApi/src/appsettings.Development.json.template [hunk 5]
```diff
@@ -1,18 +1,26 @@
 //copy this file as "appsettings.Development.json",  remove this comment and update the settings below accordingly
```
[the remark this answers is not in the document]
Comment (pritipath) [reply src/CarbonAware.WebApi/src/appsettings.Development.json.template:1]: Updated the appsettings.template in CLI to use the right config schema. 
We dont anticipate major configuration schema changes any time soon, so for now we think it is ok to have 2 copies of the template file.

**Review by esgraham — APPROVED** [review]

**Review by juzuluag — APPROVED** [review]

**Comment by pritipath** [comment]
> Currently assuming this one is **not** ready for review from the GSF - let me know if that's incorrect.

I believe all the comments were addressed, so this is ready for GSF review. @gfmatthews could you please review again to see if this looks good to you?

**Review by gfmatthews — APPROVED** [review]
New changes look good! Thank you!

**Comment by gfmatthews** [comment]
@vaughanknight - this one is now ready for review

**Review by vaughanknight — APPROVED** [review]
This looks great.  Big thank you to the team.

**Pull request merged**